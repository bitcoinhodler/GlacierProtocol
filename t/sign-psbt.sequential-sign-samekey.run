#!/bin/bash
set -e

# Test error handling when I try to sign a partial_signature PSBT with
# that same key a second time.

$GLACIERSCRIPT --regtest=$1 sign-psbt << INPUT 2>&1
y
y
y
y
y
y
../../t/sign-psbt.sequential-sign-serial.psbt
y
1
93VDtjySSuPjgLYyCR6BBTqCuaqu6X728ZfkNEDbgAdVURRd1rS
y
INPUT

# Ugly: need to delete wallet, otherwise the second signing will still have
# the first key and will create two signatures.
echo First glacierscript run complete. Deleting wallet...

# Decode the generated transaction-0*.pngs to get the PSBT for the
# second signature.
zbarimg --set "*.enable=0" --set qr.enable=1 --quiet --raw transaction-01.png | tr -d '\n' > part1.psbt
zbarimg --set "*.enable=0" --set qr.enable=1 --quiet --raw transaction-02.png >> part1.psbt

# Combine this (possibly stripped-down) partial-sig PSBT with the original PSBT
# to get a partial-sig PSBT that GlacierScript can then complete.
bitcoin-cli -regtest -rpcport=$1 -datadir=../bitcoin-data/$1 combinepsbt '["'`cat ../../t/sign-psbt.sequential-sign-serial.psbt`'", "'`cat part1.psbt`'"]' > part2.psbt


bitcoin-cli -regtest -rpcport=$1 -datadir=../bitcoin-data/$1 stop
# So ugly, but without this, I get timeouts restarting bitcoind:
sleep 5
rm -rf ../bitcoin-data/$1/regtest/wallets

echo Running glacierscript again for the second signature...

set +e

$GLACIERSCRIPT --regtest=$1 sign-psbt << INPUT 2>&1
y
y
y
y
y
y
part2.psbt
y
1
93VDtjySSuPjgLYyCR6BBTqCuaqu6X728ZfkNEDbgAdVURRd1rS
INPUT

retVal=$?
if [ $retVal -eq 0 ]; then
    echo "Error: expected glacierscript.py to fail"
    exit 1
fi
