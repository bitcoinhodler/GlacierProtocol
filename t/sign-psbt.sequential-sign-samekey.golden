Are you running this on a computer WITHOUT a network connection of any kind? (y/n)?Have the wireless cards in this computer been physically removed? (y/n)?Are you running on battery power? (y/n)?Are you running on an operating system booted from a USB drive? (y/n)?Is your screen hidden from view of windows, cameras, and other people? (y/n)?Are smartphones and all other nearby devices turned off and in a Faraday bag? (y/n)?Input a filename located in the current directory which contains the PSBT:
0.07230000 BTC in unspent inputs from cold storage address 2MyCc3oEVwSYzgmspKTjZ6HZrzeg8ajdmKR
0.07226753 BTC going to destination address bcrt1qrwxllgvtc0ns624dlped894kkfq608jmye2k50
Fee amount: 0.00003247
Confirm? (y/n): 
How many private keys will you be signing this transaction with (at least 2 required to complete transaction)?
#: Output will be a partially signed bitcoin transaction (PSBT).
Key #1: 
Is this data correct?
*** WARNING: Incorrect data may lead to loss of funds ***

0.07230000 BTC in unspent inputs from cold storage address 2MyCc3oEVwSYzgmspKTjZ6HZrzeg8ajdmKR
0.07226753 BTC going to destination address bcrt1qrwxllgvtc0ns624dlped894kkfq608jmye2k50
Fee amount: 0.00003247

Signing with private keys: 
93VDtjySSuPjgLYyCR6BBTqCuaqu6X728ZfkNEDbgAdVURRd1rS


Confirm? (y/n): 
Calculating transaction...

Incomplete PSBT (base64):
cHNidP8BAP1IAQIAAAAHJ+8Sih5dFuylcN7ZWJUt19ETcTBYg/5/rMbFcKqh5MUBAAAAAP////+HT8K90SrCVmquDo1QmNIij9ykOR7uNxKl55en0Nw8AQEAAAAA/////5TR6ZT3lHMIvxY4FNIp3nSmIcsip72pJw8x7mJ4dmDxAQAAAAD/////294NNeD1/E0SJ9dWqYLc34wHw9RMJpUAjrKpTpA2Y5AAAAAAAP////8H69iK+gLt2E97jWZNduKkN9YInjHWoEi+r3aqSi6bmAEAAAAA/////z0B9F3XKYxI0daagQlF1y+Kjwk9Lvh+VOPSdomkKvqwAAAAAAD/////GUcJHjR4u8Zstwx4fOijBWSq4mpyGnPL6EVrBMu0T7wAAAAAAP////8BgUVuAAAAAAAWABQbjf+hi8PnDSqt+HLTlrayQaeeWwAAAAAAAQBzAgAAAAEVUeLMHFhhohW4D44feKU/shjFdM7T3QDnql3VV54PUwEAAAAA/////wJC5BkAAAAAABepFD1mh6NwoEmE3MYfnU4ulVuRR+Fvh0BCDwAAAAAAF6kUQVHsDJUyftyhxMXTjQHgDZ7JPVeHAAAAAEICBP8YihMe55kApM2+aMXqhK1zuKz4QU5YlfbHsOR3nvUNObo2mjTNP3IWh3dpopLM9SwmU8eZoTgyG/jB4DqRIFlHMEQCIDBzGhgJ5kegj3/nfkO66svwRr9Fvl84auYp8f6jsh/xAiAHjq+Kf6jX3IQexPLePf6jFkVHOsjZ9RhbTZ1m0XvZSwEBBP0LAVJBBJgTe2Ijqfo5Xhb+bkmn9Kc3Ap1Eduw1hEqEaeNY1kG4WlvzVdVwaZ1jzeh9odRQ4ZZtl+5hqZtRmVoR636g9nRBBP2aafleMcO40sl5JeDxUd0cZ/O3Hjum248g2UTm6U2bXjrUwSpfm9bSE3z2mwFfljfeYar64p/fzPjvWkzYHjdBBP8YihMe55kApM2+aMXqhK1zuKz4QU5YlfbHsOR3nvUNObo2mjTNP3IWh3dpopLM9SwmU8eZoTgyG/jB4DqRIFlBBO6vwGIQCGYqmXe5U+5/evh5gYMPVhysUDhyEcddWLy3UUFPyavgz0oopARIBszVSz+AupkODv8Ea7G5MJaLlzpUrgABAIoCAAAAAbYUODn7GvWTO5sDSWj5ZX2D5IuFnKUU/ufO2E8T1KzfAQAAABcWABSfzgUdxebupXY8doh3jn/MkDyBxP////8ChmByAAAAAAAXqRRKrM2ZenvPPRzE7efl+6ZfopKfG4dAQg8AAAAAABepFEFR7AyVMn7cocTF040B4A2eyT1XhwAAAABCAgT/GIoTHueZAKTNvmjF6oStc7is+EFOWJX2x7Dkd571DTm6Npo0zT9yFod3aaKSzPUsJlPHmaE4Mhv4weA6kSBZRzBEAiApHqWTCKldDNWSBjZ0UsArooyLIdSauUDWZ1NSkoUKgAIgQqXHj6vbSjtJxBeiIDsk+kU78MX8fcPjtVUIOxX7qooBAQT9CwFSQQSYE3tiI6n6OV4W/m5Jp/SnNwKdRHbsNYRKhGnjWNZBuFpb81XVcGmdY83ofaHUUOGWbZfuYambUZlaEet+oPZ0QQT9mmn5XjHDuNLJeSXg8VHdHGfztx47ptuPINlE5ulNm1461MEqX5vW0hN89psBX5Y33mGq+uKf38z471pM2B43QQT/GIoTHueZAKTNvmjF6oStc7is+EFOWJX2x7Dkd571DTm6Npo0zT9yFod3aaKSzPUsJlPHmaE4Mhv4weA6kSBZQQTur8BiEAhmKpl3uVPuf3r4eYGDD1YcrFA4chHHXVi8t1FBT8mr4M9KKKQESAbM1Us/gLqZDg7/BGuxuTCWi5c6VK4AAQBzAgAAAAG2dx8QIdxJrRAjT0f1OGjW3mz1AKkbtwSmGymWNW7inAEAAAAA/////wKQxMABAAAAABepFBZjVHdD37Iru5GQF1qIG/7mDGv3h1BpDwAAAAAAF6kUQVHsDJUyftyhxMXTjQHgDZ7JPVeHAAAAAEICBP8YihMe55kApM2+aMXqhK1zuKz4QU5YlfbHsOR3nvUNObo2mjTNP3IWh3dpopLM9SwmU8eZoTgyG/jB4DqRIFlHMEQCIE3baQkwOlpkylFC0XUJC2MdV7uHwV0DPU0FSoDIQeLvAiAoNcb3oIRxnEaF7N9pHa2o1dIlni9tKkirH9HkaFvfXgEBBP0LAVJBBJgTe2Ijqfo5Xhb+bkmn9Kc3Ap1Eduw1hEqEaeNY1kG4WlvzVdVwaZ1jzeh9odRQ4ZZtl+5hqZtRmVoR636g9nRBBP2aafleMcO40sl5JeDxUd0cZ/O3Hjum248g2UTm6U2bXjrUwSpfm9bSE3z2mwFfljfeYar64p/fzPjvWkzYHjdBBP8YihMe55kApM2+aMXqhK1zuKz4QU5YlfbHsOR3nvUNObo2mjTNP3IWh3dpopLM9SwmU8eZoTgyG/jB4DqRIFlBBO6vwGIQCGYqmXe5U+5/evh5gYMPVhysUDhyEcddWLy3UUFPyavgz0oopARIBszVSz+AupkODv8Ea7G5MJaLlzpUrgABAIoCAAAAAUVrs+lW/IFiDoknGG60soVkRroOAZr6+v7qfw2EhDHWAQAAABcWABSxNGyrKC2N3WNy56xcUbEQfgK/of////8CgN4PAAAAAAAXqRRBUewMlTJ+3KHExdONAeANnsk9V4dggWIAAAAAABepFJogWwgjhUx/JhS6FwGfrjN22NQBhwAAAABCAgT/GIoTHueZAKTNvmjF6oStc7is+EFOWJX2x7Dkd571DTm6Npo0zT9yFod3aaKSzPUsJlPHmaE4Mhv4weA6kSBZRzBEAiBXjtaIL+uSDFB+q7y29lL6lxAjIrcC5bobiKDgQ4alUAIgE3YVgjejBCVf9goVlpnIKQEIDp/Odfe1nVtSig6TGKIBAQT9CwFSQQSYE3tiI6n6OV4W/m5Jp/SnNwKdRHbsNYRKhGnjWNZBuFpb81XVcGmdY83ofaHUUOGWbZfuYambUZlaEet+oPZ0QQT9mmn5XjHDuNLJeSXg8VHdHGfztx47ptuPINlE5ulNm1461MEqX5vW0hN89psBX5Y33mGq+uKf38z471pM2B43QQT/GIoTHueZAKTNvmjF6oStc7is+EFOWJX2x7Dkd571DTm6Npo0zT9yFod3aaKSzPUsJlPHmaE4Mhv4weA6kSBZQQTur8BiEAhmKpl3uVPuf3r4eYGDD1YcrFA4chHHXVi8t1FBT8mr4M9KKKQESAbM1Us/gLqZDg7/BGuxuTCWi5c6VK4AAQCKAgAAAAHf0xN+QsSFP7lIctYQbcrNYFw14G3KlNhvmJIIr9Xv7AEAAAAXFgAUnzEI76F1rnBBypXkwXNpy5b0oQD/////Aip7UgAAAAAAF6kUO/sbNGP63G0YdRKHk3eSWb4bfW2HkAUQAAAAAAAXqRRBUewMlTJ+3KHExdONAeANnsk9V4cAAAAAQgIE/xiKEx7nmQCkzb5oxeqErXO4rPhBTliV9sew5Hee9Q05ujaaNM0/chaHd2miksz1LCZTx5mhODIb+MHgOpEgWUcwRAIgVWzTpW4pFOmrbasDwiw8+nFAB40QutMzU5nHPgYBV0ACIFy2oS6AL6uutGLH3Cl03VuJBz7BySFz0E7w844UbwJ4AQEE/QsBUkEEmBN7YiOp+jleFv5uSaf0pzcCnUR27DWESoRp41jWQbhaW/NV1XBpnWPN6H2h1FDhlm2X7mGpm1GZWhHrfqD2dEEE/Zpp+V4xw7jSyXkl4PFR3Rxn87ceO6bbjyDZRObpTZteOtTBKl+b1tITfPabAV+WN95hqvrin9/M+O9aTNgeN0EE/xiKEx7nmQCkzb5oxeqErXO4rPhBTliV9sew5Hee9Q05ujaaNM0/chaHd2miksz1LCZTx5mhODIb+MHgOpEgWUEE7q/AYhAIZiqZd7lT7n96+HmBgw9WHKxQOHIRx11YvLdRQU/Jq+DPSiikBEgGzNVLP4C6mQ4O/wRrsbkwlouXOlSuAAEAigIAAAABEvf72yTk+oBlG5o4dhggPuCkHp02amcgrFKL7eiygLYBAAAAFxYAFMQ5dSibpN3wTFwkloHUdFHiqR97/////wKgLBAAAAAAABepFEFR7AyVMn7cocTF040B4A2eyT1Xh0qXsAEAAAAAF6kUeCtK1sAz5Cw9fm1GzLdE1+3BiCOHAAAAAEICBP8YihMe55kApM2+aMXqhK1zuKz4QU5YlfbHsOR3nvUNObo2mjTNP3IWh3dpopLM9SwmU8eZoTgyG/jB4DqRIFlHMEQCIAJrGHTE/ZXcOzLzkyyPb3sRajOAFRgWjDbvunhg3YUDAiBHQT4pHdj+Nn+VyXLy52VqnOtpHUrXcMjanXF9j/L/MwEBBP0LAVJBBJgTe2Ijqfo5Xhb+bkmn9Kc3Ap1Eduw1hEqEaeNY1kG4WlvzVdVwaZ1jzeh9odRQ4ZZtl+5hqZtRmVoR636g9nRBBP2aafleMcO40sl5JeDxUd0cZ/O3Hjum248g2UTm6U2bXjrUwSpfm9bSE3z2mwFfljfeYar64p/fzPjvWkzYHjdBBP8YihMe55kApM2+aMXqhK1zuKz4QU5YlfbHsOR3nvUNObo2mjTNP3IWh3dpopLM9SwmU8eZoTgyG/jB4DqRIFlBBO6vwGIQCGYqmXe5U+5/evh5gYMPVhysUDhyEcddWLy3UUFPyavgz0oopARIBszVSz+AupkODv8Ea7G5MJaLlzpUrgABAIoCAAAAAcH/pRDe+aRpvw3yCeNWQAZl6rnDgZrM0gIyR3WwOYmwAQAAABcWABQ5hUDm9ads73Z/M5K14RyVeBMq/P////8CsFMQAAAAAAAXqRRBUewMlTJ+3KHExdONAeANnsk9V4f0QqABAAAAABepFDONcp/aK0eqh/du5eXxLeGTYmYKhwAAAABCAgT/GIoTHueZAKTNvmjF6oStc7is+EFOWJX2x7Dkd571DTm6Npo0zT9yFod3aaKSzPUsJlPHmaE4Mhv4weA6kSBZRzBEAiASpGgxInBmjtKOVlh4v9npO7L5l3PNkJmDVVoLZ2CE0AIgKc+DU7OUS9sPL+A+gyVVotXiLEDk7bBO4Fum/ccHPTwBAQT9CwFSQQSYE3tiI6n6OV4W/m5Jp/SnNwKdRHbsNYRKhGnjWNZBuFpb81XVcGmdY83ofaHUUOGWbZfuYambUZlaEet+oPZ0QQT9mmn5XjHDuNLJeSXg8VHdHGfztx47ptuPINlE5ulNm1461MEqX5vW0hN89psBX5Y33mGq+uKf38z471pM2B43QQT/GIoTHueZAKTNvmjF6oStc7is+EFOWJX2x7Dkd571DTm6Npo0zT9yFod3aaKSzPUsJlPHmaE4Mhv4weA6kSBZQQTur8BiEAhmKpl3uVPuf3r4eYGDD1YcrFA4chHHXVi8t1FBT8mr4M9KKKQESAbM1Us/gLqZDg7/BGuxuTCWi5c6VK4AAA==

Transaction fingerprint (md5):
0d6f3a3251ec20bb57e57231793675f1
QR code for transaction written to transaction-01.png,transaction-02.png
First glacierscript run complete. Deleting wallet...
Bitcoin Core stopping
Running glacierscript again for the second signature...
Are you running this on a computer WITHOUT a network connection of any kind? (y/n)?Have the wireless cards in this computer been physically removed? (y/n)?Are you running on battery power? (y/n)?Are you running on an operating system booted from a USB drive? (y/n)?Is your screen hidden from view of windows, cameras, and other people? (y/n)?Are smartphones and all other nearby devices turned off and in a Faraday bag? (y/n)?Input a filename located in the current directory which contains the PSBT:
0.07230000 BTC in unspent inputs from cold storage address 2MyCc3oEVwSYzgmspKTjZ6HZrzeg8ajdmKR
0.07226753 BTC going to destination address bcrt1qrwxllgvtc0ns624dlped894kkfq608jmye2k50
Fee amount: 0.00003247
Confirm? (y/n): 
How many private keys will you be signing this transaction with (at least 1 required to complete transaction)?
#: Key #1: ERROR: this PSBT has already been signed by that key. Exiting...
